x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -309,12 +308,6 @@ private void createApiKeyAndIndexIt(
         Set<RoleDescriptor> userRoleDescriptors,
         ActionListener<CreateApiKeyResponse> listener
     ) {
-        if (TcpTransport.isUntrustedRemoteClusterEnabled()
```

Comment:
We have to be a little more restrictive here (and for the update): similar to how we handle [creating regular roles](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java#L224-L234) with remote access indices, we need to check that the min version across all nodes can parse remote indices. 

Otherwise, we might end up in a situation where a new node stores an API key with a remote index privilege in the security index, and another old node attempts to read it but fails.

I think a simple min version check here should work. There is a potential follow up for us around the `superuser` role around a concern I raised here: https://github.com/elastic/elasticsearch/pull/93540#pullrequestreview-1288760347

I'm leaning towards still failing the request even in the `superuser` and other reserved role cases but I have to think through it further.

Reply:
That's a very good point. The min version check makes sense for the specified roles in the request, but it might be an issue for the reserved roles - since the user has no influence over them. Would it makes sense to strip down remote index privileges from reserved user role descriptors when we detect that we are in a mixed version cluster? I haven't thought how complex it would be, but it does feel like it would cause less issues during upgrade.

Reply:
Hmm I think there may be an issue with failing the request for reserved roles. 

In cloud, API keys are used as session cookies: https://github.com/elastic/cloud/blob/8c8ff3e028f552247230b8a46c1a999c6b98d515/wiki/features/sso/request-flows.md#log-in

The keys are created against the security cluster; if we fail creation requests for users with e.g., the `superuser` role during a security cluster upgrade, this may lead to end-users being unable to login into the cloud API (until the upgrade is complete). 

Note that this is only an issue for the limited-by role descriptors. For assigned role descriptors, we can safely fail the request since there it means that an end-user is intentionally trying to create an API key for remote access and it's fair enough to return an error so that they know to retry once the upgrade is done.

For limited-by role descriptors, I see three options for dealing with the mixed cluster situation, i.e., if an API key is updated or created by a user with a role that has remote index privileges when a cluster still has old nodes:

1. fail requests, and accept the above issue 
2. fail requests, except for reserved roles. for reserved roles, strip out remote index privileges 
3. don't fail requests. for all roles, strip out remote index privileges 

I think we should avoid unnecessary glitches for end-users so currently I'm not in favor of option 1. At the same time, from an end-user perspective it's clearer to have an explicit error message. For users that don't have reserved roles (which should be the majority), we can get this with option 2. As such, I'm entertaining option 2 even it's more complex and a mix of both worlds.

WDYT?

@ywangd this is a little hairy so was hoping to get your input on this as well. Let's walk through this during our sync tomorrow.

Reply:
@slobodanadamovic sorry, only saw your comment after I posted mine. It sounds like we're both leaning towards option 2. It complicates the code but I think it might be worth it. For those `superuser` users that create a key during the upgrade, it seems quite unlikely that they do this with the intention of using it for remote access. They can furthermore "fix" the key by using the update API key API, once the upgrade is complete.

Either way, it only makes sense to address this concern once https://github.com/elastic/elasticsearch/pull/93540 is merged. 

I'll review the rest of the PR in the meantime.

Reply:
> 2. fail requests, except for reserved roles. for reserved roles, strip out remote index privileges

I was thinking the same. This option sounds more fitting.

> 3. don't fail requests. for all roles, strip out remote index privileges

If I'm not mistaken, in the mixed cluster, this option would be equivalent to the option 2 since we prevent creating roles with remote index privileges. Meaning, the only roles which can have remote index privileges in mixed version cluster are the reserved ones.

Reply:
> Meaning, the only roles which can have remote index privileges in mixed version cluster are the reserved ones.

Indeed 👍

Reply:
> the only roles which can have remote index privileges in mixed version cluster are the reserved ones.

I think this is not true. As is, in a mixed cluster, we can load such roles from either the file role store or a custom role store. 

I prefer option 3 and I'd push it even further, i.e. If assigned role descriptors do not have remote indices privileges, remove any remote indices prvileges from the limited-by role descriptors _regardless_ of the min node version. That is, even if the cluster is fully up-to-date, we still remove the remote indices privileges. I am thinking the following scenario:
1. User has a working CCS setup using the legacy model and search from QC is done using API keys
2. QC gets fully upgraded, but the existing CCS setup remains as is, i.e. user does *not* plan to migrate the existing setup and they decide to just add new ones (and maybe remove the existing ones once the new ones are working).
3. The owner user of the API keys in step 1 gets assigned a new role that contains remote privileges because they are going to use the same user for both existing and new CCS setups. 
4. Now any new API keys created by the above user will fail to work with the existing CCS setup because the legacy FC does not understand the remote privilieges (from the owning user) even if the API key itself does not have any remote privileges.

I feel this is a bad situation because it otherwise seems to be reasonable migration plan for end-users.

Reply:
We synced in person and decided to:
 - remove remote index privileges (before creating/updating API key) in a mixed cluster
 - if assigned role descriptors do not have remote indices privileges, remove any remote indices privileges from the limited-by role descriptors regardless of the min node version
 - [change `writeTo` method](https://github.com/elastic/elasticsearch/pull/93414/files#diff-62bd78e38f18d3977579107c5749cd34144482de03876461881b33e4bd74258bL358-L365) (of `RoleDescriptor` class) to skip serialization of `RemoteIndicesPrivileges` instead of failing the request when talking to a cluster on a version that does not support remote indices

Reply:
Good catch!

> Having thought about this, I want to propose one more option to deal with this issue:
>  - extend `Authentication.maybeRewriteForOlderVersion` to remove indices privileges from role descriptors (assigned and limited-by) if the target cluster is old
> - on creation and update, don't remove remote indices from limited-by role descriptors, unless we're in a mixed cluster

I like this suggestion as it's simpler and less quirky. I'd prefer to change this in this PR. This way we won't introduce logic that has to be removed.

Reply:
Sounds good! Thank you for iterating on this with me, it's a bit gnarly to untangle all the ways in which this can break 😅

Reply:
I've implemented the proposed change. It would be nice if you could take another look.